### PR TITLE
feat: Checkout action

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/workflow/execute_node_job_run.go [hunk 1]
```diff
@@ -203,49 +204,60 @@ func LoadNodeJobRunKeys(db gorp.SqlExecutor, store cache.Store, job *sdk.Workflo
...
 				Value: k.KeyID,
 			})
+
+			unBase64, err64 := base64.StdEncoding.DecodeString(k.Private)
+			if err64 != nil {
+				return nil, nil, sdk.WrapError(err64, "LoadNodeJobRunSecrets> Cannot app decode key %s", k.Name)
+			}
+			decrypted, errD := secret.Decrypt([]byte(unBase64))
+			if errD != nil {
+				log.Error("LoadNodeJobRunSecrets> Unable to decrypt app private key %s/%s: %v", n.Context.Application.Name, k.Name, errD)
```
Comment (bnjjj) [inline engine/api/workflow/execute_node_job_run.go]: LoadNodeJobRunKeys

File: engine/api/workflow/execute_node_job_run.go [hunk 2]
```diff
@@ -203,49 +204,60 @@ func LoadNodeJobRunKeys(db gorp.SqlExecutor, store cache.Store, job *sdk.Workflo
...
 			params = append(params, sdk.Parameter{
-				Name:  "cds.app." + k.Name + ".id",
+				Name:  "cds.key." + k.Name + ".id",
 				Type:  "string",
 				Value: k.KeyID,
 			})
+
+			unBase64, err64 := base64.StdEncoding.DecodeString(k.Private)
+			if err64 != nil {
+				return nil, nil, sdk.WrapError(err64, "LoadNodeJobRunSecrets> Cannot app decode key %s", k.Name)
```
Comment (bnjjj) [inline engine/api/workflow/execute_node_job_run.go]: LoadNodeJobRunKeys

File: engine/api/workflow/execute_node_job_run.go [hunk 3]
```diff
@@ -203,49 +204,60 @@ func LoadNodeJobRunKeys(db gorp.SqlExecutor, store cache.Store, job *sdk.Workflo
...
 			params = append(params, sdk.Parameter{
-				Name:  "cds.env." + k.Name + ".id",
+				Name:  "cds.key." + k.Name + ".id",
 				Type:  "string",
 				Value: k.KeyID,
 			})
+
+			unBase64, err64 := base64.StdEncoding.DecodeString(k.Private)
+			if err64 != nil {
+				return nil, nil, sdk.WrapError(err64, "LoadNodeJobRunSecrets> Cannot decode env key %s", k.Name)
```
Comment (bnjjj) [inline engine/api/workflow/execute_node_job_run.go]: LoadNodeJobRunKeys

File: engine/api/workflow/execute_node_job_run.go [hunk 4]
```diff
@@ -203,49 +204,60 @@ func LoadNodeJobRunKeys(db gorp.SqlExecutor, store cache.Store, job *sdk.Workflo
...
+			}
+			decrypted, errD := secret.Decrypt([]byte(unBase64))
+			if errD != nil {
+				log.Error("LoadNodeJobRunSecrets> Unable to decrypt env private key %s/%s: %v", n.Context.Environment.Name, k.Name, errD)
```
Comment (bnjjj) [inline engine/api/workflow/execute_node_job_run.go]: LoadNodeJobRunKeys

File: engine/worker/builtin_gittag.go [hunk 5]
```diff
@@ -39,127 +32,66 @@ func runGitTag(w *currentWorker) BuiltInAction {
...
 
-		//Get the key
-		key, errK := vcs.GetSSHKey(*params, keysDirectory, authPrivateKey)
-		if errK != nil && errK != sdk.ErrKeyNotFound {
+		v, errT := semver.Make(tagName.Value)
+		if errT != nil {
 			res := sdk.Result{
 				Status: sdk.StatusFail.String(),
-				Reason: fmt.Sprintf("Unable to setup ssh key. %s", errK),
+				Reason: fmt.Sprintf("Tag name is not semver compatible"),
```
Comment (bnjjj) [inline engine/worker/builtin_gittag.go]: useless fmt.Sprintf

File: engine/worker/builtin_gittag.go [hunk 6]
```diff
@@ -39,127 +32,66 @@ func runGitTag(w *currentWorker) BuiltInAction {
...
-			msg = tagMessage.Value
-		}
-
-		v, errT := semver.Make(tagName.Value)
-		if errT != nil {
+		if userTag == "" {
 			res := sdk.Result{
 				Status: sdk.StatusFail.String(),
-				Reason: fmt.Sprintf("Tag name is not semver compatible"),
+				Reason: fmt.Sprintf("No user find to perform tag"),
```
Comment (bnjjj) [inline engine/worker/builtin_gittag.go]: useless fmt.Sprintf

File: engine/worker/vcs_strategy.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"fmt"
+	"github.com/ovh/cds/sdk"
```
Comment (bnjjj) [inline engine/worker/vcs_strategy.go]: space

File: ui/src/assets/i18n/fr.json [hunk 8]
```diff
@@ -554,6 +555,14 @@
   "user_edit_title": "Édition de l'utisateur",
   "user_none": "Aucun utilisateur",
 
+  "vcs_connection" : "Connexion : ",
+  "vcs_user" : "Utilisateur : ",
+  "vcs_password" : "Mot de passe : ",
+  "vcs_branch" : "Branche : ",
+  "vcs_default_branch" : "Branche par defaut : ",
```
Comment (bnjjj) [inline ui/src/assets/i18n/fr.json]: défaut

File: engine/api/action/builtin.go [hunk 9]
```diff
@@ -95,40 +95,30 @@ The public key have to be granted on your repository`,
 		return err
 	}
 
+	// ----------------------------------- Checkout Application    -----------------------
+	checkoutApplication := sdk.NewAction(sdk.CheckoutApplication)
+	checkoutApplication.Type = sdk.BuiltinAction
+	checkoutApplication.Description = `CDS Builtin Action.
+Clone a repository into a new directory.`
```
Comment (fsamin) [inline engine/api/action/builtin.go]: s/Clone/Checkout

File: engine/api/api_routes.go [hunk 10]
```diff
@@ -113,6 +113,7 @@ func (api *API) InitRouter() {
 	r.Handle("/project/{permProjectKey}/variable/{name}/audit", r.GET(api.getVariableAuditInProjectHandler))
 	r.Handle("/project/{permProjectKey}/applications", r.GET(api.getApplicationsHandler), r.POST(api.addApplicationHandler))
 	r.Handle("/project/{permProjectKey}/notifications", r.GET(api.getProjectNotificationsHandler))
+	r.Handle("/project/{permProjectKey}/all/keys", r.GET(api.getAllKeysProjectHandler))
```
Comment (fsamin) [inline engine/api/api_routes.go]: i don't like this route

File: engine/api/application/application_exporter.go [hunk 11]
```diff
@@ -16,13 +16,14 @@ import (
...
 	app, errload := LoadByName(db, cache, key, appName, u,
-		LoadOptions.WithVariablesWithClearPassword,
-		LoadOptions.WithKeys,
+		LoadOptions.WithVariablesWithClearPassword, LoadOptions.WithClearKeys,
 	)
 	if errload != nil {
 		return 0, sdk.WrapError(errload, "application.Export> Cannot load application %s", appName)
 	}
 
+	// load clear key
```
Comment (fsamin) [inline engine/api/application/application_exporter.go]: remove

File: engine/api/application/gorp_model.go [hunk 12]
```diff
@@ -46,7 +58,13 @@ func (a *dbApplication) PostUpdate(db gorp.SqlExecutor) error {
 	if err != nil {
 		return err
 	}
-	if _, err := db.Exec("update application set metadata = $2 where id = $1", a.ID, b); err != nil {
+
+	v, err := json.Marshal(a.RepositoryStrategy)
```
Comment (fsamin) [inline engine/api/application/gorp_model.go]: you should encrypt this, because it can contains a password

File: sdk/application.go [hunk 13]
```diff
@@ -40,6 +41,17 @@ type Application struct {
...
+// RepositoryStrategy represent the way to use the repository
+type RepositoryStrategy struct {
+	ConnectionType string `json:"connection_type"`
+	SSHKey         string `json:"ssh_key"`
+	User           string `json:"user"`
+	Password       string `json:"password"`
+	Branch         string `json:"branch"`
+	DefaultBranch  string `json:"default_branch"`
+	PGPKey         string `json:"pgp_key"`
+}
```
Comment (fsamin) [inline sdk/application.go]: recursive ftw

Comment (sguiheux) [reply sdk/application.go]: next iteration

File: sdk/exportentities/application.go [hunk 14]
```diff
@@ -60,5 +68,14 @@ func NewApplication(app sdk.Application, withPermissions bool, keys []EncryptedK
 			Value: e.Content,
 		}
 	}
+
+	a.VCSBranch = app.RepositoryStrategy.Branch
+	a.VCSConnectionType = app.RepositoryStrategy.ConnectionType
```
Comment (fsamin) [inline sdk/exportentities/application.go]: if type is https, a.VCSConnectionType must be empty

File: engine/api/application/application_parser.go [hunk 15]
```diff
@@ -86,6 +86,29 @@ func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, e
...
+		app.RepositoryStrategy.ConnectionType = "https"
+	}
+	if eapp.VCSPassword != "" {
+		clearPWD, err := decryptFunc(db, proj.ID, eapp.VCSPassword)
+		if err != nil {
+			return nil, sdk.WrapError(err, "ParseAndImport> Unable to decrypt vcs password")
+		}
+		app.RepositoryStrategy.Password = clearPWD
+		if errE := EncryptVCSStrategyPassword(app); errE != nil {
+			return nil, sdk.WrapError(err, "ParseAndImport> Cannot encrypt vcs password")
```
Comment (yesnault) [inline engine/api/application/application_parser.go]: errE

**Pull request merged**